internal/ociartifact/store.go

```diff
@@ -0,0 +1,432 @@
+type Store struct {
+	rootPath      string
+	systemContext *types.SystemContext
+	impl          Impl
+}
+
+// NewStore creates a new OCI artifact store.
+func NewStore(rootPath string, systemContext *types.SystemContext) *Store {
+	return &Store{
+		rootPath:      filepath.Join(rootPath, "artifacts"),
+		systemContext: systemContext,
+		impl:          &defaultImpl{},
+	}
+}
+
+// PullOptions can be used to customize the pull behavior.
+type PullOptions struct {
+	// EnforceConfigMediaType can be set to enforce a specific manifest config
+	// media type.
+	EnforceConfigMediaType string
+
+	// MaxSize is the maximum size of the artifact to be allowed to stay
+	// in-memory. This is only useful when requesting the artifact data using
+	// PullData.
+	// Will be set to a default of 1MiB if not specified (zero) or below zero.
+	MaxSize uint64
+
+	// CopyOptions are the copy options passed down to libimage.
+	CopyOptions *libimage.CopyOptions
+}
+
+// PullData downloads the artifact into the local storage and returns its data.
+// Returns ErrNotFound if the artifact is not available.
+func (s *Store) PullData(ctx context.Context, ref string, opts *PullOptions) ([]ArtifactData, error) {
+	opts = sanitizeOptions(opts)
+
+	log.Infof(ctx, "Pulling OCI artifact from ref: %s", ref)
+
+	name, err := s.impl.ParseNormalizedNamed(ref)
+	if err != nil {
+		return nil, fmt.Errorf("parse image name: %w", err)
+	}
+
+	name = reference.TagNameOnly(name) // make sure to add ":latest" if needed
+
+	dockerRef, err := s.impl.DockerNewReference(name)
+	if err != nil {
+		return nil, fmt.Errorf("create docker reference: %w", err)
+	}
+
+	manifestBytes, err := s.PullManifest(ctx, dockerRef, opts)
+	if err != nil {
+		return nil, fmt.Errorf("pull artifact: %w", err)
+	}
+
+	artifactData, err := s.artifactData(ctx, digest.FromBytes(manifestBytes).Encoded(), opts.MaxSize)
+	if err != nil {
+		return nil, fmt.Errorf("get artifact data: %w", err)
+	}
+
+	return artifactData, nil
+}
+
+// PullManifest tries to pull the artifact and returns the manifest bytes if the
+// provided reference is a valid OCI artifact.
+//
+// Returns ErrIsAnImage if the artifact is a container image.
+//
+// enforceConfigMediaType can be used to allow only a certain config media type.
+// copyOptions will be passed down to libimage.
+func (s *Store) PullManifest(
+	ctx context.Context,
+	ref types.ImageReference,
+	opts *PullOptions,
+) (manifestBytes []byte, err error) {
+	opts = sanitizeOptions(opts)
+	strRef := s.impl.DockerReferenceString(ref)
+
+	log.Debugf(ctx, "Checking if source reference is an OCI artifact: %v", strRef)
+
+	src, err := s.impl.NewImageSource(ctx, ref, s.systemContext)
+	if err != nil {
+		return nil, fmt.Errorf("build image source: %w", err)
+	}
+
+	defer func() {
+		if err := s.impl.CloseImageSource(src); err != nil {
+			log.Warnf(ctx, "Unable to close image source: %v", err)
+		}
+	}()
+
+	manifestBytes, mimeType, err := s.impl.GetManifest(ctx, src, nil)
```

Comment:
I was wondering if you could accommodate https://github.com/containers/skopeo/pull/2527/files#diff-6cac070631a328cbecde69cd12534e6b77254bcf5d366c4210afaa76e70bfe67R109 in this commit.
This might fixes https://github.com/cri-o/cri-o/issues/9028

Reply:
That's already in: https://github.com/cri-o/cri-o/pull/8996/files#diff-214891d73adc1da245487a6f80db7c6891c760310da9179d4c912b46047d6848R68